documentation/debug.md

```diff
@@ -141,10 +142,163 @@ gvd --debugger $HOME/devel/toolchains/gcc-linaro-arm-none-eabi-4.9-2014.09_linux
+  tee_session_ioctl                    11330    54380860 us     4799.722 us     35403.79 us
+  tee_session_invoke_be                11330    54330744 us     4795.299 us     162939.5 us
+  tz_invoke                            11330    54014297 us     4767.369 us     573472.7 us
+  tee_ioctl                             1139    2893849 us     2540.692 us     2841179 us
+  tee_session_create_fd                 1135    2889859 us     2546.131 us     2615175 us
+  ...
+```
+
+#### 2.2.2.1 Oneliners
+```
+# Print also the core number in the log
+for core in `seq 0 7`; do echo core: $core; cat trace_stat/function$core; done
+```
+
+```
+# The functions that are called mostly:
+cat trace_stat/function0  | sort -nk2 -r | less
+```
+
+```
+# The functions taking most time:
+cat trace_stat/function0  | sort -nk5 -r | less
+```
+
+### 2.2.3 Using the function_graph
+The function_graph will give you the call flow and also tell you the amount of
+time spent in the functions. There are ways to turn of sleep time and not count
+time spent when calling other functions. Let us say that your are interested in
+knowing how much various open, invoke and close and the call_tee command takes,
+then you can do like this:
+
+```
+echo "tz_open" > set_ftrace_filter
+echo "tz_close" >> set_ftrace_filter
+echo "tz_invoke" >> set_ftrace_filter
+echo "call_tee*" >> set_ftrace_filter
+
+# Don't count the time if you are being schduled out
+echo 0 > options/sleep-time
+
+# Enable the function_graph tracer
+echo "function_graph" > current_tracer
+```
+
+Now if you run `xtest` and then done, read the contents of trace, you will see
+something like this:
+```
+# CPU  DURATION                  FUNCTION CALLS
+# |     |   |                     |   |   |   |
+ 2)               |  tz_open [optee_armtz]() {
+ 2) ! 3145.834 us |    call_tee.isra.13 [optee_armtz]();
+ 2) ! 3222.500 us |  }
+ 2)               |  tz_invoke [optee_armtz]() {
+ 2) ! 125.833 us  |    call_tee.isra.13 [optee_armtz]();
+ 2) ! 166.667 us  |  }
+ 2)               |  tz_invoke [optee_armtz]() {
+ 2) ! 135.833 us  |    call_tee.isra.13 [optee_armtz]();
+ 2) ! 170.833 us  |  }
+ 2)               |  tz_invoke [optee_armtz]() {
+ 2) ! 153.334 us  |    call_tee.isra.13 [optee_armtz]();
+ 2) ! 186.667 us  |  }
+...
+```
+
+### 2.2.4 Options
+If you don't want to count the time when being scheduled out, then run:
+```
+echo 0 > options/sleep-time
+```
+
+If you only want to measure the time spent *in* the function, then disable the
+graph-time.
+```
+echo 0 > options/graph-time
+```
+
+# 3. Known issues
 1. Printing the call stack using `bt` makes gdb go into an endless loop.
    Temporary workaround, in gdb, instead of simply writing `bt`, also mention
    how many frames you would like to see, for example `bt 10`.
 2. Cannot set breakpoints when the system is up and running. Workaround, set the
-   breakpoints before booting up the system.
+   breakpoints befora booting up the system.
```

Comment:
I've spotted this error, will revert it when applying tags or addressing other comments.